File: frontend/src/lib/documents.ts
/**
 * Knowledge Library API client (M7.3).
 *
 * Mirrors `backend/api/routers/documents.py`:
 *
 *   POST   /api/documents/ingest           multipart OR JSON
 *   GET    /api/documents
 *   GET    /api/documents/{doc_id}
 *   GET    /api/documents/{doc_id}/chunks
 *   POST   /api/documents/{doc_id}:reindex
 *   DELETE /api/documents/{doc_id}
 *   POST   /api/documents/search
 *
 * The ingest helper has two flavours so callers can drop a file *or*
 * paste raw markdown without thinking about content-type negotiation.
 */
import { api } from "@/lib/api";
import type {
  DocChunkHit,
  DocumentChunkPage,
  DocumentListResponse,
  DocumentSearchResponse,
  DocumentSourceKind,
  IngestDocumentJSONInput,
  IngestDocumentResponse,
  KnowledgeDocument,
  UpdateDocumentInput,
} from "@/types/api";

function qs(params: Record<string, unknown>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null || v === "") continue;
    search.set(k, String(v));
  }
  const s = search.toString();
  return s ? `?${s}` : "";
}

const enc = encodeURIComponent;

export interface ListDocumentsParams {
  user_id?: string;
  tag?: string;
  limit?: number;
  offset?: number;
}

export interface IngestDocumentFileArgs {
  file: File;
  title?: string;
  tags?: string[];
  source_kind?: DocumentSourceKind;
  source_uri?: string;
  target_tokens?: number;
  overlap_tokens?: number;
}

export const documentsApi = {
  list(params: ListDocumentsParams = {}): Promise<DocumentListResponse> {
    return api<DocumentListResponse>(
      `/api/documents${qs(params as Record<string, unknown>)}`,
    );
  },

  get(docId: string): Promise<KnowledgeDocument> {
    return api<KnowledgeDocument>(`/api/documents/${enc(docId)}`);
  },

  listChunks(
    docId: string,
    params: { offset?: number; limit?: number } = {},
  ): Promise<DocumentChunkPage> {
    return api<DocumentChunkPage>(
      `/api/documents/${enc(docId)}/chunks${qs(params as Record<string, unknown>)}`,
    );
  },

  ingestJSON(payload: IngestDocumentJSONInput): Promise<IngestDocumentResponse> {
    return api<IngestDocumentResponse>("/api/documents/ingest", {
      method: "POST",
      json: payload,
    });
  },

  ingestFile(args: IngestDocumentFileArgs): Promise<IngestDocumentResponse> {
    const fd = new FormData();
    fd.append("file", args.file);
    if (args.title !== undefined) fd.append("title", args.title);
    if (args.tags && args.tags.length > 0) fd.append("tags", args.tags.join(", "));
    if (args.source_kind) fd.append("source_kind", args.source_kind);
    if (args.source_uri) fd.append("source_uri", args.source_uri);
    if (args.target_tokens !== undefined)
      fd.append("target_tokens", String(args.target_tokens));
    if (args.overlap_tokens !== undefined)
      fd.append("overlap_tokens", String(args.overlap_tokens));
    return api<IngestDocumentResponse>("/api/documents/ingest", {
      method: "POST",
      body: fd,
    });
  },

  reindex(
    docId: string,
    params: { target_tokens?: number; overlap_tokens?: number } = {},
  ): Promise<IngestDocumentResponse> {
    return api<IngestDocumentResponse>(`/api/documents/${enc(docId)}:reindex`, {
      method: "POST",
      json: {
        target_tokens: params.target_tokens ?? 800,
        overlap_tokens: params.overlap_tokens ?? 100,
      },
    });
  },

  delete(docId: string): Promise<void> {
    return api<void>(`/api/documents/${enc(docId)}`, { method: "DELETE" });
  },

  /**
   * P14.B — partial metadata edit. Five fields at most. Does NOT
   * re-chunk or re-embed; for body changes go through ``reindex``
   * (which re-chunks + re-embeds the whole doc).
   *
   * Title changes propagate to chunk-level vector metadata server-side
   * (the chunks denormalise ``doc_title`` for cheap search rendering).
   */
  patchMetadata(docId: string, body: UpdateDocumentInput): Promise<KnowledgeDocument> {
    return api<KnowledgeDocument>(`/api/documents/${enc(docId)}`, {
      method: "PATCH",
      json: body,
    });
  },

  search(
    q: string,
    params: { top_k?: number; filters?: Record<string, unknown> } = {},
  ): Promise<DocumentSearchResponse> {
    return api<DocumentSearchResponse>("/api/documents/search", {
      method: "POST",
      json: {
        q,
        top_k: params.top_k ?? 5,
        filters: params.filters ?? {},
      },
    });
  },
};

export type { DocChunkHit };

File: sdk/python/aaf/documents.py
"""``/api/documents/*`` sub-client (M7.3 — Knowledge Library / RAG).

Mirrors :mod:`backend.api.routers.documents`. Both async and sync facades
share the same surface; either accept a typed payload or a raw ``dict``
for callers who'd rather not import the DTOs.
"""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Any

from .models import (
    DocChunk,
    DocChunkHit,
    IngestDocumentResponse,
    KnowledgeDocument,
)

if TYPE_CHECKING:  # pragma: no cover
    from .client import AAFClient, AsyncAAFClient


def _qs(params: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in params.items() if v not in (None, "", [])}


def _multipart_fields(
    *,
    title: str | None,
    tags: list[str] | None,
    source_kind: str | None,
    source_uri: str | None,
    target_tokens: int | None,
    overlap_tokens: int | None,
) -> dict[str, str]:
    out: dict[str, str] = {}
    if title is not None:
        out["title"] = title
    if tags:
        out["tags"] = ", ".join(tags)
    if source_kind:
        out["source_kind"] = source_kind
    if source_uri:
        out["source_uri"] = source_uri
    if target_tokens is not None:
        out["target_tokens"] = str(target_tokens)
    if overlap_tokens is not None:
        out["overlap_tokens"] = str(overlap_tokens)
    return out


# ---------------------------------------------------------------------------
# Async
# ---------------------------------------------------------------------------


class AsyncDocumentsAPI:
    def __init__(self, client: AsyncAAFClient) -> None:
        self._client = client

    async def list_all(
        self,
        *,
        user_id: str | None = None,
        tag: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[KnowledgeDocument]:
        body = await self._client.request_json(
            "GET",
            "/api/documents",
            params=_qs(
                {"user_id": user_id, "tag": tag, "limit": limit, "offset": offset}
            ),
        )
        return [KnowledgeDocument.model_validate(it) for it in (body or {}).get("items", [])]

    async def get(self, doc_id: str) -> KnowledgeDocument:
        body = await self._client.request_json("GET", f"/api/documents/{doc_id}")
        return KnowledgeDocument.model_validate(body)

    async def get_chunks(
        self, doc_id: str, *, offset: int = 0, limit: int = 100
    ) -> list[DocChunk]:
        body = await self._client.request_json(
            "GET",
            f"/api/documents/{doc_id}/chunks",
            params={"offset": offset, "limit": limit},
        )
        return [DocChunk.model_validate(it) for it in (body or {}).get("items", [])]

    async def ingest_text(
        self,
        *,
        title: str = "",
        raw_text: str,
        source_kind: str = "note",
        source_uri: str = "",
        summary: str = "",
        tags: list[str] | None = None,
        target_tokens: int = 800,
        overlap_tokens: int = 100,
    ) -> IngestDocumentResponse:
        payload: dict[str, Any] = {
            "title": title,
            "raw_text": raw_text,
            "source_kind": source_kind,
            "source_uri": source_uri,
            "summary": summary,
            "tags": list(tags or []),
            "target_tokens": target_tokens,
            "overlap_tokens": overlap_tokens,
        }
        body = await self._client.request_json(
            "POST", "/api/documents/ingest", json_body=payload
        )
        return IngestDocumentResponse.model_validate(body)

    async def ingest_file(
        self,
        path: str | Path,
        *,
        title: str | None = None,
        tags: list[str] | None = None,
        source_kind: str | None = None,
        source_uri: str | None = None,
        target_tokens: int = 800,
        overlap_tokens: int = 100,
    ) -> IngestDocumentResponse:
        p = Path(path)
        files = {"file": (p.name, p.read_bytes(), _guess_mime(p))}
        body = await self._client.request_json(
            "POST",
            "/api/documents/ingest",
            files=files,
            data=_multipart_fields(
                title=title,
                tags=tags,
                source_kind=source_kind,
                source_uri=source_uri,
                target_tokens=target_tokens,
                overlap_tokens=overlap_tokens,
            ),
        )
        return IngestDocumentResponse.model_validate(body)

    async def reindex(
        self, doc_id: str, *, target_tokens: int = 800, overlap_tokens: int = 100
    ) -> IngestDocumentResponse:
        body = await self._client.request_json(
            "POST",
            f"/api/documents/{doc_id}:reindex",
            json_body={"target_tokens": target_tokens, "overlap_tokens": overlap_tokens},
        )
        return IngestDocumentResponse.model_validate(body)

    async def delete(self, doc_id: str) -> None:
        await self._client.request_json("DELETE", f"/api/documents/{doc_id}")

    async def search(
        self,
        q: str,
        *,
        top_k: int = 5,
        filters: dict[str, Any] | None = None,
    ) -> list[DocChunkHit]:
        body = await self._client.request_json(
            "POST",
            "/api/documents/search",
            json_body={"q": q, "top_k": top_k, "filters": filters or {}},
        )
        return [DocChunkHit.model_validate(it) for it in (body or {}).get("items", [])]


# ---------------------------------------------------------------------------
# Sync
# ---------------------------------------------------------------------------


class DocumentsAPI:
    def __init__(self, client: AAFClient) -> None:
        self._client = client

    def list_all(
        self,
        *,
        user_id: str | None = None,
        tag: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> list[KnowledgeDocument]:
        body = self._client.request_json(
            "GET",
            "/api/documents",
            params=_qs(
                {"user_id": user_id, "tag": tag, "limit": limit, "offset": offset}
            ),
        )
        return [KnowledgeDocument.model_validate(it) for it in (body or {}).get("items", [])]

    def get(self, doc_id: str) -> KnowledgeDocument:
        body = self._client.request_json("GET", f"/api/documents/{doc_id}")
        return KnowledgeDocument.model_validate(body)

    def get_chunks(
        self, doc_id: str, *, offset: int = 0, limit: int = 100
    ) -> list[DocChunk]:
        body = self._client.request_json(
            "GET",
            f"/api/documents/{doc_id}/chunks",
            params={"offset": offset, "limit": limit},
        )
        return [DocChunk.model_validate(it) for it in (body or {}).get("items", [])]

    def ingest_text(
        self,
        *,
        title: str = "",
        raw_text: str,
        source_kind: str = "note",
        source_uri: str = "",
        summary: str = "",
        tags: list[str] | None = None,
        target_tokens: int = 800,
        overlap_tokens: int = 100,
    ) -> IngestDocumentResponse:
        body = self._client.request_json(
            "POST",
            "/api/documents/ingest",
            json_body={
                "title": title,
                "raw_text": raw_text,
                "source_kind": source_kind,
                "source_uri": source_uri,
                "summary": summary,
                "tags": list(tags or []),
                "target_tokens": target_tokens,
                "overlap_tokens": overlap_tokens,
            },
        )
        return IngestDocumentResponse.model_validate(body)

    def ingest_file(
        self,
        path: str | Path,
        *,
        title: str | None = None,
        tags: list[str] | None = None,
        source_kind: str | None = None,
        source_uri: str | None = None,
        target_tokens: int = 800,
        overlap_tokens: int = 100,
    ) -> IngestDocumentResponse:
        p = Path(path)
        files = {"file": (p.name, p.read_bytes(), _guess_mime(p))}
        body = self._client.request_json(
            "POST",
            "/api/documents/ingest",
            files=files,
            data=_multipart_fields(
                title=title,
                tags=tags,
                source_kind=source_kind,
                source_uri=source_uri,
                target_tokens=target_tokens,
                overlap_tokens=overlap_tokens,
            ),
        )
        return IngestDocumentResponse.model_validate(body)

    def reindex(
        self, doc_id: str, *, target_tokens: int = 800, overlap_tokens: int = 100
    ) -> IngestDocumentResponse:
        body = self._client.request_json(
            "POST",
            f"/api/documents/{doc_id}:reindex",
            json_body={"target_tokens": target_tokens, "overlap_tokens": overlap_tokens},
        )
        return IngestDocumentResponse.model_validate(body)

    def delete(self, doc_id: str) -> None:
        self._client.request_json("DELETE", f"/api/documents/{doc_id}")

    def search(
        self,
        q: str,
        *,
        top_k: int = 5,
        filters: dict[str, Any] | None = None,
    ) -> list[DocChunkHit]:
        body = self._client.request_json(
            "POST",
            "/api/documents/search",
            json_body={"q": q, "top_k": top_k, "filters": filters or {}},
        )
        return [DocChunkHit.model_validate(it) for it in (body or {}).get("items", [])]


def _guess_mime(path: Path) -> str:
    suffix = path.suffix.lower()
    if suffix == ".pdf":
        return "application/pdf"
    if suffix in {".md", ".markdown"}:
        return "text/markdown"
    return "text/plain"


__all__ = ["AsyncDocumentsAPI", "DocumentsAPI"]
